### PR TITLE
[web] Use greatest span font size on the parent paragraph

### PR DESCRIPTION
**Pull request by mdebbar**
The default font size (14px) is applied to the `<body>` of the page, and the `<p>` elements inherit the font size from the body. When the spans have a font size smaller than than the default font size, they get vertically misaligned.

Because the default font size is only 14px, the misalignment isn't big enough to make a visible difference in screenshots. For that reason, the test is setting a font size of `100px` on the body to exacerbate the issue and make it clearly visible in the screenshots.

Here's the before and after:

|**Before**|**After**|
|---|---|
|before|after|

Screenshots: https://github.com/flutter/goldens/pull/188

Fixes https://github.com/flutter/flutter/issues/79901

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart [hunk 1]
```diff
@@ -134,6 +134,7 @@ class CanvasParagraph implements EngineParagraph {
 
     // 1. Set paragraph-level styles.
     _applyNecessaryParagraphStyles(element: rootElement, style: paragraphStyle);
+    _applySpanStylesToParagraph(element: rootElement, spans: spans);
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:137]: what if you have only placeholder/widgetspans?

**Comment by ditman** [comment]
Is this going to mess up with the performance of rendering rich text or similar? Paragraphs-with-many-spans?

What about *not* flipping the renderer from canvas to html in the cases that the original issue demonstrate? When there's no hover, there's no problem; however on hover the rendering of those paragraphs flips to DOM...

File: lib/web_ui/lib/src/engine/text/canvas_paragraph.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/text/canvas_paragraph.dart:137]: Addressed with test.

**Review by ferhatb — APPROVED** [review]

**Comment by mdebbar** [comment]
@ditman switching between canvas <-> dom depends on many things. There could be other reasons that caused the switch to dom (e.g. images), and in that case, it would be cheaper to continue using dom instead of switching back to canvas.

There are certainly cases where a paragraph forces rendering on canvas or dom (e.g. paragraphs that contain decorations always use dom today). But ideally, we should reduce those cases to the minimum, so the majority of paragraphs can be rendered anywhere.

In this specific case, the looping happens only once per paragraph. Once the DOM element is created, it will be cached and cloned for future renders.

**Pull request merged**